Indigo shade related fixes and other changes

PepperCode1 opened the pull request:
- Apply disabled shade from vanilla quads directly to material. `MutableQuadViewImpl#fromVanilla` now modifies the passed material to force disable diffuse if the vanilla quad has shade disabled, which allows the material to accurately represent the diffuse state of the quad. `QuadViewImpl` no longer stores a hidden `shade` boolean that cannot be copied by a FRAPI user. These changes do not change the behavior of any Indigo methods.
- Fix `AoCalculator` using AO face data computed with a potentially different shade state. With enhanced AO mode, after a non-axis-aligned quad had AO data calculated and cached, any future quads on the same block with a different shade state would be rendered as though they had the shade state of the first quad. The AO data cache now separates data by shade state to fix this issue.
- Respect cull parameter passed to non-terrain rendering.
- Other minor cleanup and performance improvements

PepperCode1 committed:
Apply disabled shade from vanilla quads directly to material

- Remove QuadViewImpl.shade

PepperCode1 committed:
Fix enhanced AO calculation and respect non-terrain culling state

- Fix AoCalculator using AO face data computed with a potentially different shade state
- Move non-cached computation code to separate method in AoCalculator
- Turn AoCalculator's brightnessFunc and aoFunc into abstract methods
- Do not check null check world in non-terrain AO calculation since it cannot be null
- Pass through lightFace and shade state as method arguments in AoCalculator methods to prevent additional lookups
- Do not check for the axis aligned flag in AbstractQuadRenderer.shadeFlatQuad
- Respect cull parameter passed to non-terrain rendering by merging TerrainBlockRenderInfo into BlockRenderInfo
- Use reusable search pos when calling Block.shouldDrawSide to prevent additional BlockPos allocation
- Change BlockRenderContext.render and TerrainRenderContext.tessellateBlock to return void since return value is no longer used
- Remove QuadViewImpl.vertexStart since it is unused

Technici4n commented on fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/mesh/MutableQuadViewImpl.java:
```diff
@@ -116,9 +117,13 @@ public final MutableQuadViewImpl fromVanilla(BakedQuad quad, RenderMaterial mate
 		nominalFace(quad.getFace());
 		colorIndex(quad.getColorIndex());
+
+		if (!quad.hasShade()) {
```
Note to self: When the quad has shade and the material does not, an "opinionated" choice needs to be made on whether to shade or not here. I'm fine with the current behavior of disabling the shade if either the quad or the material disables it.

Technici4n commented on fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/render/BlockRenderInfo.java:
```diff
@@ -84,7 +94,24 @@ int blockColor(int colorIndex) {
 	}
 
 	boolean shouldDrawFace(Direction face) {
```
Mark it as `@Nullable` maybe?

Technici4n commented on fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/render/AbstractQuadRenderer.java:
```diff
@@ -235,9 +234,9 @@ int flatBrightness(MutableQuadViewImpl quad, BlockState blockState, BlockPos pos
 	private void shadeFlatQuad(MutableQuadViewImpl quad) {
-		if ((quad.geometryFlags() & GeometryHelper.AXIS_ALIGNED_FLAG) == 0 || quad.hasVertexNormals()) {
-			// Quads that aren't direction-aligned or that have vertex normals need to be shaded
-			// using interpolation - vanilla can't handle them. Generally only applies to modded models.
+		if (quad.hasVertexNormals()) {
+			// Quads that have vertex normals need to be shaded using interpolation - vanilla can't
+			// handle them. Generally only applies to modded models.
```
Why?

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/impl/renderer/RendererHolder.java:
```diff
@@ -16,17 +16,30 @@
+	@Nullable
+	private static Renderer activeRenderer = null;
+
+	/** avoids null test every call to {@link #isPresent()}. */
```
That's weird, `null` checks should largely be fast enough. But it was already there so /shrug.

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/Renderer.java:
```diff
@@ -23,13 +23,41 @@
+	 * and materials. Will return null if no render plug in is active.
+	 */
+	@Nullable
+	static Renderer get() {
```
Could we maybe have `getNullable` that has the current behavior and `get` that throws an exception if there is no registered renderer?

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java:
```diff
@@ -18,32 +18,43 @@
 	 * <p>Only one {@link Renderer} plug-in can be active in any game instance.
 	 * If a second mod attempts to register this method will throw an UnsupportedOperationException.
+	 *
+	 * <p>Use {@link Renderer#register(Renderer)} instead.
```
```suggestion
	 * @deprecated Use {@link Renderer#register(Renderer)} instead.
```

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java:
```diff
@@ -18,32 +18,43 @@
 	 * Access to the current {@link Renderer} for creating and retrieving model builders
 	 * and materials. Will return null if no render plug in is active.
+	 *
+	 * <p>Use {@link Renderer#get()} instead.
```
```suggestion
	 * @deprecated Use {@link Renderer#get()} instead.
```

Technici4n commented on fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java:
```diff
@@ -18,32 +18,43 @@
 	/**
 	 * Performant test for {@link #getRenderer()} != null.
+	 *
+	 * <p>Use {@link Renderer#isPresent()} instead.
```
```suggestion
	 * @deprecated Use {@link Renderer#isPresent()} instead.
```

Technici4n reviewed (changes requested):
Looks very good overall, just a few comments.

PepperCode1 committed:
Add suggestions

- Mark Direction parameter to BlockRenderInfo.shouldDrawFace as Nullable
- Reuse MaterialFinder in FrameBakedModel

modmuss50 commented:
Note to future self: Backport to 1.18.2 and 1.19.2
